python/utils/OpaqueArguments.h

```diff
@@ -156,47 +182,33 @@ packArgs(OpaqueArguments &argData, py::args args,
     auto kernelArgTy = kernelFuncOp.getArgument(i).getType();
     llvm::TypeSwitch<mlir::Type, void>(kernelArgTy)
         .Case([&](mlir::ComplexType ty) {
-          if (!py::hasattr(arg, "real"))
-            throw std::runtime_error("invalid complex element type");
-          if (!py::hasattr(arg, "imag"))
-            throw std::runtime_error("invalid complex element type");
-
-          std::complex<double> *ourAllocatedArg = new std::complex<double>();
+          checkArgumentType<py_ext::Complex>(arg, i);
           if (isa<Float64Type>(ty.getElementType())) {
-            *ourAllocatedArg =
-                std::complex<double>(PyFloat_AsDouble(arg.attr("real").ptr()),
-                                     PyFloat_AsDouble(arg.attr("imag").ptr()));
+            std::complex<double> *ourAllocatedArg = new std::complex<double>();
+            *ourAllocatedArg = arg.cast<std::complex<double>>();
 
             argData.emplace_back(ourAllocatedArg, [](void *ptr) {
               delete static_cast<std::complex<double> *>(ptr);
             });
           } else {
-            *ourAllocatedArg = std::complex<float>(
-                arg.attr("real").cast<float>(), arg.attr("imag").cast<float>());
+            std::complex<float> *ourAllocatedArg = new std::complex<float>();
+            *ourAllocatedArg = arg.cast<std::complex<float>>();
 
             argData.emplace_back(ourAllocatedArg, [](void *ptr) {
               delete static_cast<std::complex<float> *>(ptr);
             });
```

Comment:
I wonder if there is a simpler way to do this pattern with the deallocations. It just seems really repetitive here...

Reply:
I'll think of something... could the fix be in a different pr?